resource-managers/kubernetes/core/src/main/scala/org/apache/spark/security/kubernetes/TokenRefreshService.scala

```diff
@@ -0,0 +1,239 @@
+          Base64.decodeBase64(data))))
+        creds
+    }
+    val tokens = credentials.map {
+      creds =>
+        creds.getAllTokens.asScala.toList
+    }
+    tokens.getOrElse(Nil)
+  }
+
+  private def renewTokens(tokens: List[Token[_ <: TokenIdentifier]])
+      : Map[Token[_ <: TokenIdentifier], Long] = {
+    tokens.map(token => {
+      val expireTime = try {
+          token.renew(hadoopConf)
+        } catch {
+          case t: Throwable =>
+            logWarning(t.getMessage, t)
+            hasError = true
+
+          getRetryTime()
+        }
+      (token, expireTime)
+    }).toMap
+  }
+
+  private def getRetryTime() = clock.nowInMillis() + RENEW_TASK_DEADLINE_LOOK_AHEAD_MILLIS
+}
+
+private class RenewTask(renew: Renew,
+                        hadoopConf: Configuration,
+                        refreshService: ActorRef,
+                        clock: Clock) extends Runnable with Logging {
+
+  private var hasError = false
+
+  override def run() : Unit = {
+    val deadline = renew.expireTime + RENEW_TASK_DEADLINE_LOOK_AHEAD_MILLIS
+    val newExpireTimeByToken : Map[Token[_ <: TokenIdentifier], Long] =
+      renew.expireTimeByToken.map {
+        item =>
+          val token = item._1
+          val expireTime = item._2
+          val newExpireTime =
+            if (expireTime <= deadline) {
+              try {
+                token.renew(hadoopConf)
+              } catch {
+                case t: Throwable =>
+                  logWarning(t.getMessage, t)
+                  hasError = true
+
+                getRetryTime()
+              }
+            } else {
+              expireTime
+            }
+          (token, newExpireTime)
+      }
+      .toMap
+    if (newExpireTimeByToken.nonEmpty) {
+      val nextExpireTime = newExpireTimeByToken.values.min
+      logInfo(s"Renewed with the result $newExpireTimeByToken. Next expire time $nextExpireTime")
+      val numConsecutiveErrors = if (hasError) renew.numConsecutiveErrors + 1 else 0
+      refreshService ! Renew(nextExpireTime, newExpireTimeByToken, renew.secret,
+        numConsecutiveErrors)
+    } else {
+      logWarning(s"Got an empty token list with ${renew.secret.getMetadata.getSelfLink}")
+    }
+  }
+
+  private def getRetryTime() = clock.nowInMillis() + RENEW_TASK_DEADLINE_LOOK_AHEAD_MILLIS
+}
+
+private class Clock {
+
+  def nowInMillis() : Long = System.currentTimeMillis()
+}
+
+private case class UpdateRefreshList(secrets: List[Secret])
+private case class StartRefresh(secret: Secret)
+private case class Renew(expireTime: Long,
+                         expireTimeByToken: Map[Token[_ <: TokenIdentifier], Long],
+                         secret: Secret,
+                         numConsecutiveErrors: Int)
+private case class StopRefresh(secret: Secret)
+
+private object TokenRefreshService {
+
+  def apply(system: ActorSystem) : ActorRef = {
+    UserGroupInformation.loginUserFromKeytab(
```

Comment:
Reminder to myself. We'll have to refresh the kerberos TGT as well periodically.